File: engine/src/nuke-detect.ts
import fs from "node:fs"
import path from "node:path"
import type { NukeInstallation } from "./engine-settings.js"

function parseVersion(dirName: string, exeName: string): string {
  const m = /Nuke(\d+(?:\.\d+)?)/i.exec(dirName) ?? /Nuke(\d+(?:\.\d+)?)/i.exec(exeName)
  return m?.[1] ?? "unknown"
}

/** Scan Windows Program Files for Nuke executables (newest first). */
export function detectNukeInstallations(): NukeInstallation[] {
  const found: NukeInstallation[] = []
  const seen = new Set<string>()
  const roots = [
    process.env["ProgramFiles"],
    process.env["ProgramFiles(x86)"],
  ].filter(Boolean) as string[]

  for (const root of roots) {
    let entries: fs.Dirent[] = []
    try {
      entries = fs.readdirSync(root, { withFileTypes: true })
    } catch {
      continue
    }
    for (const entry of entries) {
      if (!entry.isDirectory() || !/^Nuke/i.test(entry.name)) continue
      const dir = path.join(root, entry.name)
      let exes: fs.Dirent[] = []
      try {
        exes = fs.readdirSync(dir, { withFileTypes: true })
      } catch {
        continue
      }
      for (const exe of exes) {
        if (!exe.isFile() || !/^Nuke\d/i.test(exe.name) || !exe.name.endsWith(".exe")) continue
        if (/crash|assist/i.test(exe.name)) continue
        const full = path.join(dir, exe.name)
        if (seen.has(full)) continue
        seen.add(full)
        const version = parseVersion(entry.name, exe.name)
        found.push({
          exePath: full.replace(/\\/g, "/"),
          version,
          label: `Nuke ${version} (${entry.name})`,
          sortKey: version,
        })
      }
    }
  }

  found.sort((a, b) => (b.sortKey ?? "").localeCompare(a.sortKey ?? "", undefined, { numeric: true }))
  return found
}

export function resolveSampleNkTemplate(engineRoot: string): string | null {
  const candidates = [
    path.join(engineRoot, "python", "templates", "review_mp4.nk"),
    path.join(engineRoot, "..", "..", "sample.nk"),
  ]
  for (const c of candidates) {
    if (fs.existsSync(c)) return c
  }
  return null
}

File: engine/python/modules/nuke_detect.py
"""
Detect Foundry Nuke installations on Windows.
"""

from __future__ import annotations

import glob
import os
import re
from pathlib import Path
from typing import Any, Dict, List, Optional


def _parse_version_from_path(exe_path: str) -> str:
    parent = Path(exe_path).parent.name
    match = re.search(r"Nuke(\d+(?:\.\d+)?)", parent, re.I)
    if match:
        return match.group(1)
    match = re.search(r"Nuke(\d+(?:\.\d+)?)", Path(exe_path).name, re.I)
    if match:
        return match.group(1)
    return "unknown"


def detect_nuke_installations() -> List[Dict[str, Any]]:
    """Return sorted list (newest first) of Nuke executables."""
    found: List[Dict[str, Any]] = []
    patterns = [
        r"C:\Program Files\Nuke*\Nuke*.exe",
        r"C:\Program Files (x86)\Nuke*\Nuke*.exe",
    ]
    env_exe = os.environ.get("NUKE_EXE") or os.environ.get("NUKE_PATH")
    if env_exe:
        p = Path(env_exe)
        if p.is_file():
            found.append(_entry(str(p)))
        elif (p / "Nuke15.1.exe").is_file():
            found.append(_entry(str(p / "Nuke15.1.exe")))

    seen = set()
    for pattern in patterns:
        for match in glob.glob(pattern):
            low = match.lower()
            if "crash" in low or "nukeassist" in low:
                continue
            if "nuke" not in Path(match).stem.lower():
                continue
            if match in seen:
                continue
            seen.add(match)
            found.append(_entry(match))

    found.sort(key=lambda x: x.get("sortKey", ""), reverse=True)
    return found


def _entry(exe_path: str) -> Dict[str, Any]:
    version = _parse_version_from_path(exe_path)
    label = f"Nuke {version} ({Path(exe_path).parent.name})"
    return {
        "exePath": exe_path.replace("\\", "/"),
        "version": version,
        "label": label,
        "sortKey": version,
    }


def resolve_sample_nk_template() -> Optional[str]:
    engine_root = Path(__file__).resolve().parent.parent.parent
    candidates = [
        engine_root / "python" / "templates" / "review_mp4.nk",
        engine_root.parent.parent / "sample.nk",
    ]
    for path in candidates:
        if path.is_file():
            return str(path)
    return None
